pkg/cluster/manager.go

```diff
@@ -1583,6 +1580,126 @@ func (m *Manager) ScaleOut(
 	return nil
 }
 
+// DestroyTombstone destroy and remove instances that is in tombstone state
+func (m *Manager) DestroyTombstone(
+	clusterName string,
+	gOpt operator.Options,
+	skipConfirm bool,
+) error {
+	var (
+		sshTimeout = gOpt.SSHTimeout
+		sshType    = gOpt.SSHType
+	)
+
+	metadata, err := m.meta(clusterName)
+	// allow specific validation errors so that user can recover a broken
+	// cluster if it is somehow in a bad state.
+	if err != nil &&
+		!errors.Is(perrs.Cause(err), spec.ErrNoTiSparkMaster) {
+		return perrs.AddStack(err)
+	}
+
+	topo := metadata.GetTopology()
+	base := metadata.GetBaseMeta()
+
+	clusterMeta := metadata.(*spec.ClusterMeta)
+	cluster := clusterMeta.Topology
+
+	if !operator.NeedCheckTombstone(cluster) {
+		return nil
+	}
+
+	tlsCfg, err := topo.TLSConfig(m.specManager.Path(clusterName, spec.TLSCertKeyDir))
+	if err != nil {
+		return err
+	}
+
+	b := task.NewBuilder().
+		SSHKeySet(
+			m.specManager.Path(clusterName, "ssh", "id_rsa"),
+			m.specManager.Path(clusterName, "ssh", "id_rsa.pub")).
+		ClusterSSH(topo, base.User, sshTimeout, sshType, metadata.GetTopology().BaseTopo().GlobalOptions.SSHType)
+
+	var nodes []string
+	b.
+		Func("FindTomestoneNodes", func(ctx *task.Context) (err error) {
+			nodes, err = operator.DestroyTombstone(ctx, cluster, true /* returnNodesOnly */, gOpt, tlsCfg)
+			if !skipConfirm {
+				err = cliutil.PromptForConfirmOrAbortError(
+					color.HiYellowString(fmt.Sprintf("Will destroy these nodes: %v\nDo you confirm this action? [y/N]:", nodes)),
+				)
+				if err != nil {
+					return err
+				}
+			}
+			log.Infof("Start destroy Tombstone nodes: %v ...", nodes)
+			return err
+		}).
+		ClusterOperate(cluster, operator.DestroyTombstoneOperation, gOpt, tlsCfg).
+		UpdateMeta(clusterName, clusterMeta, nodes).
+		UpdateTopology(clusterName, m.specManager.Path(clusterName), clusterMeta, nodes)
+
+	var regenConfigTasks []task.Task
+	deletedNodes := set.NewStringSet(nodes...)
+	topo.IterInstance(func(instance spec.Instance) {
```

Comment:
Can we add a function such as `RefreshConfig` to build the `regenConfigTasks`? I found that there is a lot of redundant code: https://github.com/pingcap/tiup/pull/958/files#diff-5609f7b865f14ce911938ab0a8f21caa07446244d9c106a5df27ba9c08cf80e6R1386

Notice: there are many other places use the same logic, we can replace that logic with just a function call: 
- https://github.com/pingcap/tiup/blob/master/pkg/cluster/manager.go#L702
- https://github.com/pingcap/tiup/blob/master/pkg/cluster/manager.go#L2073

Reply:
Good idea

Reply:
As this PR contains 200+ lines modified, IMO, I'll trigger another PR to do the refactor work. BTW the file `manager.go`  is too large(~ 3000 loc),  I want to split this file into some small files If you agree with it. :)

Reply:
It's very nice if you can split the `manager.go`, I want to do that days ago.

I think we can add a package (eg. `tiup/pkg/cluster/manager`) which contains several .go files, each of them has different function (eg. split by `Deploy`, `Start`, 'ScaleOut` .etc.)